marimo/_runtime/runtime.py

```diff
@@ -845,6 +851,13 @@ def set_cell_config(self, request: SetCellConfigRequest) -> None:
                 dataflow.transitive_closure(self.graph, cells_to_run)
             )
 
+    def set_app_metadata(self, request: UpdateAppMetadataRequest) -> None:
+        """Update app metadata."""
+        self.app_metadata = request.metadata
+
+        # Set __file__ to the globals
+        self.globals["__file__"] = self.app_metadata.filename
```

Comment:
Don't overwrite `__file__` if the user has defined it (double underscore variables are not local). 

```suggestion
        if "__file__" not in self.graph.definitions:
            self.globals["__file__"] = self.app_metadata.filename
```

Reply:
On the other hand, then the frontend would need to register a rename as a run, and the backend needs to send a completed run when  it's done renaming, which is kind of weird.

After giving it some thought, I vote for only writing to __file__ at kernel creation time, to not violate our guarantee that program state and outputs are consistent. I think it's simpler, matches CPython, and good enough for most use cases.